core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java

```diff
@@ -430,18 +439,28 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
 
         if (evt instanceof ProxyConnectionEvent) {
             proxyDestinationAddress = ((ProxyConnectionEvent) evt).destinationAddress();
-            if (protocol != null) {
-                // SessionProtocol event was already triggered.
-                tryCompleteSessionPromise(ctx);
-            }
+            tryCompleteSessionPromise(ctx);
             return;
         }
 
         logger.warn("{} Unexpected user event: {}", channel, evt);
     }
 
     private void tryCompleteSessionPromise(ChannelHandlerContext ctx) {
-        if (!sessionPromise.trySuccess(channel)) {
+        if (protocol == null || !isChannelActive) {
+            return;
+        }
+        if (poolKey.proxyConfig.proxyType() != ProxyType.DIRECT && proxyDestinationAddress == null) {
+            // ProxyConnectionEvent is necessary for a proxied connection.
+            return;
+        }
+        if (protocol.isExplicitHttp2() && !isSettingsFrameReceived) {
+            // Http2Settings should be received for HTTP/2.
+            return;
+        }
+
+        sessionTimeoutFuture.cancel(false);
+        if (!sessionPromise.trySuccess(channel) && !sessionPromise.isSuccess()) {
```

Comment:
Question: Is there a case that `!sessionPromise.trySuccess(channel) == true` and `!sessionPromise.isSuccess() == false`?

Reply:
It is possible if `sessionPromise` has been completed exceptionally already. Are you asking when `sessionPromise` fails?

Reply:
Nop, I've asked it if we can remove `!sessionPromise.isSuccess()` condition or not. 😉

Reply:
It is necessary in case `tryCompleteSessionPromise()` is called after `sessionPromise` is completed successfully. Let me add a comment and flip the `if` condition to understand easily.